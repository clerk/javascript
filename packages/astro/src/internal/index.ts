/**
 * The following code will be used in order to be injected as script via the astro integration.
 * F.e.
 *
 * injectScript('before-hydration', `...`)
 */

import { createClerkInstance } from './create-clerk-instance';
import { createInjectionScriptRunner } from './create-injection-script-runner';

const runInjectionScript = createInjectionScriptRunner(createClerkInstance);

export { runInjectionScript };

export { generateSafeId } from './utils/generateSafeId';
