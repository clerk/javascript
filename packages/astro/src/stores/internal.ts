import type {
  ActiveSessionResource,
  Clerk,
  ClientResource,
  InitialState,
  OrganizationResource,
  UserResource,
} from '@clerk/types';
import { atom, map } from 'nanostores';

export const $csrState = map<{
  isLoaded: boolean;
  client: ClientResource | undefined | null;
  user: UserResource | undefined | null;
  session: ActiveSessionResource | undefined | null;
  organization: OrganizationResource | undefined | null;
}>({
  isLoaded: false,
  client: null,
  user: null,
  session: null,
  organization: null,
});

export const $initialState = map<InitialState>();

// Use atom instead of `map` to prohibit key changes and allow only replacing the whole object
export const $clerk = atom<Clerk | null>(null);
