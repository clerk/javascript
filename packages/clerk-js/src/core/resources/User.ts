import type {
  BackupCodeJSON,
  BackupCodeResource,
  CreateEmailAddressParams,
  CreateExternalAccountParams,
  CreatePhoneNumberParams,
  CreateWeb3WalletParams,
  DeletedObjectJSON,
  DeletedObjectResource,
  EmailAddressResource,
  ExternalAccountJSON,
  ExternalAccountResource,
  GetUserOrganizationInvitationsParams,
  ImageResource,
  OrganizationMembershipResource,
  PhoneNumberResource,
  RemoveUserPasswordParams,
  SamlAccountResource,
  SetProfileImageParams,
  TOTPJSON,
  TOTPResource,
  UpdateUserParams,
  UpdateUserPasswordParams,
  UserJSON,
  UserResource,
  VerifyTOTPParams,
  Web3WalletResource,
} from '@clerk/types';

import { unixEpochToDate } from '../../utils/date';
import { normalizeUnsafeMetadata } from '../../utils/resourceParams';
import { getFullName } from '../../utils/user';
import { BackupCode } from './BackupCode';
import type { RetrieveMembershipsParams } from './internal';
import {
  BaseResource,
  DeletedObject,
  EmailAddress,
  ExternalAccount,
  Image,
  OrganizationMembership,
  PhoneNumber,
  SamlAccount,
  SessionWithActivities,
  TOTP,
  UserOrganizationInvitation,
  Web3Wallet,
} from './internal';

export class User extends BaseResource implements UserResource {
  pathRoot = '/me';

  id = '';
  externalId: string | null = null;
  username: string | null = null;
  emailAddresses: EmailAddressResource[] = [];
  phoneNumbers: PhoneNumberResource[] = [];
  web3Wallets: Web3WalletResource[] = [];
  externalAccounts: ExternalAccountResource[] = [];

  /**
   * @experimental
   */
  samlAccounts: SamlAccountResource[] = [];

  organizationMemberships: OrganizationMembershipResource[] = [];
  passwordEnabled = false;
  firstName: string | null = null;
  lastName: string | null = null;
  fullName: string | null = null;
  primaryEmailAddressId: string | null = null;
  primaryEmailAddress: EmailAddressResource | null = null;
  primaryPhoneNumberId: string | null = null;
  primaryPhoneNumber: PhoneNumberResource | null = null;
  primaryWeb3WalletId: string | null = null;
  primaryWeb3Wallet: Web3WalletResource | null = null;
  /**
   * @deprecated  Use `imageUrl` instead.
   */
  profileImageUrl = '';
  imageUrl = '';
  twoFactorEnabled = false;
  totpEnabled = false;
  backupCodeEnabled = false;
  publicMetadata: UserPublicMetadata = {};
  unsafeMetadata: UserUnsafeMetadata = {};
  createOrganizationEnabled = false;
  deleteSelfEnabled = false;
  lastSignInAt: Date | null = null;
  updatedAt: Date | null = null;
  createdAt: Date | null = null;

  private cachedSessionsWithActivities: SessionWithActivities[] | null = null;

  static isUserResource(resource: unknown): resource is User {
    return !!resource && resource instanceof User;
  }

  constructor(data: UserJSON) {
    super();
    this.fromJSON(data);
  }

  protected path(): string {
    return this.pathRoot;
  }

  isPrimaryIdentification = (ident: EmailAddressResource | PhoneNumberResource | Web3WalletResource): boolean => {
    switch (ident.constructor) {
      case EmailAddress:
        return this.primaryEmailAddressId === ident.id;
      case PhoneNumber:
        return this.primaryPhoneNumberId === ident.id;
      case Web3Wallet:
        return this.primaryWeb3WalletId === ident.id;
      default:
        return false;
    }
  };

  createEmailAddress = (params: CreateEmailAddressParams): Promise<EmailAddressResource> => {
    const { email } = params || {};
    return new EmailAddress(
      {
        email_address: email,
      },
      this.path() + '/email_addresses/',
    ).create();
  };

  createPhoneNumber = (params: CreatePhoneNumberParams): Promise<PhoneNumberResource> => {
    const { phoneNumber } = params || {};
    return new PhoneNumber(
      {
        phone_number: phoneNumber,
      },
      this.path() + '/phone_numbers/',
    ).create();
  };

  createWeb3Wallet = (params: CreateWeb3WalletParams): Promise<Web3WalletResource> => {
    const { web3Wallet } = params || {};
    return new Web3Wallet(
      {
        web3_wallet: web3Wallet,
      },
      this.path() + '/web3_wallets/',
    ).create();
  };

  createExternalAccount = async (params: CreateExternalAccountParams): Promise<ExternalAccountResource> => {
    const { strategy, redirectUrl, additionalScopes, redirect_url } = params || {};

    const json = (
      await BaseResource._fetch<ExternalAccountJSON>({
        path: '/me/external_accounts',
        method: 'POST',
        body: {
          strategy,
          redirect_url: redirectUrl || redirect_url,
          additional_scope: additionalScopes,
        } as any,
      })
    )?.response as unknown as ExternalAccountJSON;

    return new ExternalAccount(json, this.path() + '/external_accounts');
  };

  createTOTP = async (): Promise<TOTPResource> => {
    const json = (
      await BaseResource._fetch<TOTPJSON>({
        path: '/me/totp',
        method: 'POST',
      })
    )?.response as unknown as TOTPJSON;

    return new TOTP(json);
  };

  verifyTOTP = async ({ code }: VerifyTOTPParams): Promise<TOTPResource> => {
    const json = (
      await BaseResource._fetch<ExternalAccountJSON>({
        path: '/me/totp/attempt_verification',
        method: 'POST',
        body: { code } as any,
      })
    )?.response as unknown as TOTPJSON;

    return new TOTP(json);
  };

  disableTOTP = async (): Promise<DeletedObjectResource> => {
    const json = (
      await BaseResource._fetch<DeletedObjectJSON>({
        path: '/me/totp',
        method: 'DELETE',
      })
    )?.response as unknown as DeletedObjectJSON;

    return new DeletedObject(json);
  };

  createBackupCode = async (): Promise<BackupCodeResource> => {
    const json = (
      await BaseResource._fetch<BackupCodeJSON>({
        path: this.path() + '/backup_codes/',
        method: 'POST',
      })
    )?.response as unknown as BackupCodeJSON;

    return new BackupCode(json);
  };

  update = (params: UpdateUserParams): Promise<UserResource> => {
    return this._basePatch({
      body: normalizeUnsafeMetadata(params),
    });
  };

  updatePassword = (params: UpdateUserPasswordParams): Promise<UserResource> => {
    return this._basePost({
      body: params,
      path: `${this.path()}/change_password`,
    });
  };

  removePassword = (params: RemoveUserPasswordParams): Promise<UserResource> => {
    return this._basePost({
      body: params,
      path: `${this.path()}/remove_password`,
    });
  };

  delete = (): Promise<void> => {
    return this._baseDelete({ path: '/me' });
  };

  getSessions = async (): Promise<SessionWithActivities[]> => {
    if (this.cachedSessionsWithActivities) {
      return this.cachedSessionsWithActivities;
    }
    const res = await SessionWithActivities.retrieve();
    this.cachedSessionsWithActivities = res;
    return res;
  };

  setProfileImage = (params: SetProfileImageParams): Promise<ImageResource> => {
    const { file } = params || {};

    if (file === null) {
      return Image.delete(`${this.path()}/profile_image`);
    }
    return Image.create(`${this.path()}/profile_image`, {
      file,
    });
  };

  getOrganizationInvitations = (params?: GetUserOrganizationInvitationsParams) => {
    return UserOrganizationInvitation.retrieve(params);
  };

  getOrganizationMemberships = async (
    retrieveMembership: RetrieveMembershipsParams,
  ): Promise<OrganizationMembership[]> => {
    return await OrganizationMembership.retrieve(retrieveMembership);
  };

  get verifiedExternalAccounts() {
    return this.externalAccounts.filter(externalAccount => externalAccount.verification?.status == 'verified');
  }

  get unverifiedExternalAccounts() {
    return this.externalAccounts.filter(externalAccount => externalAccount.verification?.status != 'verified');
  }

  get hasVerifiedEmailAddress() {
    return this.emailAddresses.filter(email => email.verification.status === 'verified').length > 0;
  }

  get hasVerifiedPhoneNumber() {
    return this.phoneNumbers.filter(phone => phone.verification.status === 'verified').length > 0;
  }

  protected fromJSON(data: UserJSON | null): this {
    if (!data) {
      return this;
    }

    this.id = data.id;
    this.externalId = data.external_id;
    this.firstName = data.first_name;
    this.lastName = data.last_name;
    if (this.firstName || this.lastName) {
      this.fullName = getFullName({ firstName: this.firstName, lastName: this.lastName });
    }

    this.profileImageUrl = data.profile_image_url;
    this.imageUrl = data.image_url;
    this.username = data.username;
    this.passwordEnabled = data.password_enabled;
    this.emailAddresses = (data.email_addresses || []).map(
      ea => new EmailAddress(ea, this.path() + '/email_addresses'),
    );

    this.primaryEmailAddressId = data.primary_email_address_id;
    this.primaryEmailAddress = this.emailAddresses.find(({ id }) => id === this.primaryEmailAddressId) || null;

    this.phoneNumbers = (data.phone_numbers || []).map(ph => new PhoneNumber(ph, this.path() + '/phone_numbers'));

    this.primaryPhoneNumberId = data.primary_phone_number_id;
    this.primaryPhoneNumber = this.phoneNumbers.find(({ id }) => id === this.primaryPhoneNumberId) || null;

    this.web3Wallets = (data.web3_wallets || []).map(ph => new Web3Wallet(ph, this.path() + '/web3_wallets'));

    this.primaryWeb3WalletId = data.primary_web3_wallet_id;
    this.primaryWeb3Wallet = this.web3Wallets.find(({ id }) => id === this.primaryWeb3WalletId) || null;

    this.externalAccounts = (data.external_accounts || []).map(
      ea => new ExternalAccount(ea, this.path() + '/external_accounts'),
    );

    this.organizationMemberships = (data.organization_memberships || []).map(om => new OrganizationMembership(om));

    this.samlAccounts = (data.saml_accounts || []).map(sa => new SamlAccount(sa, this.path() + '/saml_accounts'));

    this.publicMetadata = data.public_metadata;
    this.unsafeMetadata = data.unsafe_metadata;

    this.totpEnabled = data.totp_enabled;
    this.backupCodeEnabled = data.backup_code_enabled;
    this.twoFactorEnabled = data.two_factor_enabled;

    this.createOrganizationEnabled = data.create_organization_enabled;
    this.deleteSelfEnabled = data.delete_self_enabled;

    if (data.last_sign_in_at) {
      this.lastSignInAt = unixEpochToDate(data.last_sign_in_at);
    }

    this.updatedAt = unixEpochToDate(data.updated_at);
    this.createdAt = unixEpochToDate(data.created_at);
    return this;
  }
}
