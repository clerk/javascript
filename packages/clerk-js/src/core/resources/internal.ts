export type { default as Clerk } from '../clerk';
export * from './Base';
export * from './UserSettings';
export * from './AuthConfig';
export * from './Client';
export * from './DeletedObject';
export * from './DisplayConfig';
export * from './EmailAddress';
export * from './Environment';
export * from './Error';
export * from './ExternalAccount';
export * from './IdentificationLink';
export * from './Image';
export * from './PhoneNumber';
export * from './Organization';
export * from './OrganizationInvitation';
export * from './OrganizationMembership';
export * from './SamlAccount';
export * from './Session';
export * from './SessionWithActivities';
export * from './SignIn';
export * from './SignUp';
export * from './Token';
export * from './TOTP';
export * from './User';
export * from './Verification';
export * from './Web3Wallet';
