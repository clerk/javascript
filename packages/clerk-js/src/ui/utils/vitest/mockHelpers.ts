// import { jest } from '@jest/globals';
import type { LoadedClerk } from '@clerk/types';
import type { ActiveSessionResource } from '@clerk/types';
import { type Mocked, vi } from 'vitest';

import type { RouteContextValue } from '../../router';

type FunctionLike = (...args: any) => any;

type DeepVitestMocked<T> = T extends FunctionLike
  ? Mocked<T>
  : T extends object
    ? {
        [k in keyof T]: DeepVitestMocked<T[k]>;
      }
    : T;

// Removing jest.Mock type for now, relying on inference
type MockMap<T = any> = {
  [K in { [K in keyof T]: T[K] extends (...args: any[]) => any ? K : never }[keyof T]]?: ReturnType<typeof vi.fn>;
};

const mockProp = <T>(obj: T, k: keyof T, mocks?: MockMap<T>) => {
  if (typeof obj[k] === 'function') {
    const mockFn = mocks?.[k] ?? vi.fn();
    (obj[k] as unknown as ReturnType<typeof vi.fn>) = mockFn;
  }
};

const mockMethodsOf = <T extends Record<string, any> | null = any>(
  obj: T,
  options?: {
    exclude: (keyof T)[];
    mocks: MockMap<T>;
  },
) => {
  if (!obj) {
    return;
  }
  Object.keys(obj)
    .filter(key => !options?.exclude.includes(key as keyof T))
    // Pass the specific MockMap for the object T being mocked
    .forEach(k => mockProp(obj, k, options?.mocks));
};

export const mockClerkMethods = (clerk: LoadedClerk): DeepVitestMocked<LoadedClerk> => {
  // Cast clerk to any to allow mocking properties
  const clerkAny = clerk as any;

  mockMethodsOf(clerkAny);
  if (clerkAny.client) {
    mockMethodsOf(clerkAny.client.signIn);
    mockMethodsOf(clerkAny.client.signUp);
    clerkAny.client.sessions?.forEach((session: ActiveSessionResource) => {
      const sessionAny = session as any;
      mockMethodsOf(sessionAny, {
        exclude: ['checkAuthorization'],
        mocks: {
          // Ensure touch mock matches expected signature if available, otherwise basic mock
          touch: vi.fn(() => Promise.resolve(session)),
        },
      });
      if (sessionAny.user) {
        mockMethodsOf(sessionAny.user);
        sessionAny.user.emailAddresses?.forEach((m: any) => mockMethodsOf(m));
        sessionAny.user.phoneNumbers?.forEach((m: any) => mockMethodsOf(m));
        sessionAny.user.externalAccounts?.forEach((m: any) => mockMethodsOf(m));
        sessionAny.user.organizationMemberships?.forEach((m: any) => {
          mockMethodsOf(m);
          if (m.organization) {
            mockMethodsOf(m.organization);
          }
        });
        sessionAny.user.passkeys?.forEach((m: any) => mockMethodsOf(m));
      }
    });
  }
  mockProp(clerkAny, 'navigate');
  mockProp(clerkAny, 'setActive');
  mockProp(clerkAny, 'redirectWithAuth');
  mockProp(clerkAny, '__internal_navigateWithError');
  return clerkAny as DeepVitestMocked<LoadedClerk>;
};

export const mockRouteContextValue = ({ queryString = '' }: Partial<DeepVitestMocked<RouteContextValue>>) => {
  return {
    basePath: '',
    startPath: '',
    flowStartPath: '',
    fullPath: '',
    indexPath: '',
    currentPath: '',
    queryString,
    queryParams: {},
    getMatchData: vi.fn(),
    matches: vi.fn(),
    baseNavigate: vi.fn(),
    navigate: vi.fn(() => Promise.resolve(true)),
    resolve: vi.fn((to: string) => new URL(to, 'https://clerk.com')),
    refresh: vi.fn(),
    params: {},
  } as RouteContextValue; // Keep original type assertion, DeepVitestMocked applied to input only
};
