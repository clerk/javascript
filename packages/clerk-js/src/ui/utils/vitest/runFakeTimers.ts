import { vi } from 'vitest';
// import { jest } from '@jest/globals';
import { act } from '@testing-library/react';

type WithAct = <T>(fn: T) => T;
const withAct = ((fn: any) =>
  (...args: any) => {
    act(() => {
      fn(...args);
    });
  }) as WithAct;

const advanceTimersByTime = withAct(vi.advanceTimersByTime.bind(vi));
const runAllTimers = withAct(vi.runAllTimers.bind(vi));
const runOnlyPendingTimers = withAct(vi.runOnlyPendingTimers.bind(vi));

const createFakeTimersHelpers = () => {
  return { advanceTimersByTime, runAllTimers, runOnlyPendingTimers };
};

type FakeTimersHelpers = ReturnType<typeof createFakeTimersHelpers>;
type RunFakeTimersCallback = (timers: FakeTimersHelpers) => void | Promise<void>;

export const runFakeTimers = <T extends RunFakeTimersCallback, R extends ReturnType<T>>(
  cb: T,
): R extends Promise<any> ? Promise<void> : void => {
  vi.useFakeTimers();
  const res = cb(createFakeTimersHelpers());
  if (res && 'then' in res) {
    // @ts-expect-error
    return res.finally(() => vi.useRealTimers());
  }
  vi.useRealTimers();
  // @ts-ignore
  return;
};
