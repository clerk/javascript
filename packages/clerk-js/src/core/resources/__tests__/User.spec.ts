import type { UserJSON } from '@clerk/types';
import { describe, expect, it, vi } from 'vitest';

import { BaseResource } from '../internal';
import { User } from '../User';

describe('User', () => {
  it('creates an external account', async () => {
    const externalAccountJSON = {
      object: 'external_account',
      provider: 'dropbox',
      verification: {
        external_verification_redirect_url: 'https://www.example.com',
      },
    };

    // @ts-ignore
    BaseResource._fetch = vi.fn().mockReturnValue(Promise.resolve({ response: externalAccountJSON }));

    const user = new User({
      email_addresses: [],
      phone_numbers: [],
      web3_wallets: [],
      external_accounts: [],
    } as unknown as UserJSON);

    await user.createExternalAccount({
      strategy: 'oauth_dropbox',
      redirectUrl: 'https://www.example.com',
      additionalScopes: ['view'],
    });

    // @ts-ignore
    expect(BaseResource._fetch).toHaveBeenCalledWith({
      method: 'POST',
      path: '/me/external_accounts',
      body: {
        redirect_url: 'https://www.example.com',
        strategy: 'oauth_dropbox',
        additional_scope: ['view'],
      },
    });
  });

  it('creates a web3 wallet', async () => {
    const targetWeb3Wallet = '0x0000000000000000000000000000000000000000';
    const web3WalletJSON = {
      object: 'web3_wallet',
      web3_wallet: targetWeb3Wallet,
    };

    // @ts-ignore
    BaseResource._fetch = vi.fn().mockReturnValue(Promise.resolve({ response: web3WalletJSON }));

    const user = new User({
      email_addresses: [],
      phone_numbers: [],
      web3_wallets: [],
      external_accounts: [],
    } as unknown as UserJSON);

    await user.createWeb3Wallet({ web3Wallet: targetWeb3Wallet });

    // @ts-ignore
    expect(BaseResource._fetch).toHaveBeenCalledWith({
      method: 'POST',
      path: '/me/web3_wallets/',
      body: {
        web3_wallet: targetWeb3Wallet,
      },
    });
  });

  it('denotes if has a verified email address', () => {
    const userWithUnverifiedEmail = new User({
      email_addresses: [
        {
          emailAddress: 'unverified@clerk.com',
          verification: null,
        },
      ],
      phone_numbers: [],
      web3_wallets: [],
      external_accounts: [],
    } as unknown as UserJSON);

    expect(userWithUnverifiedEmail.hasVerifiedEmailAddress).toEqual(false);

    const userWithVerifiedEmail = new User({
      email_addresses: [
        {
          emailAddress: 'unverified@clerk.com',
          verification: {
            status: 'verified',
          },
        },
      ],
      phone_numbers: [],
      web3_wallets: [],
      external_accounts: [],
    } as unknown as UserJSON);

    expect(userWithVerifiedEmail.hasVerifiedEmailAddress).toEqual(true);
  });

  it('denotes if has a verified phone number', () => {
    const userWithUnverifiedPhone = new User({
      email_addresses: [],
      phone_numbers: [
        {
          phoneNumber: '+306900000000',
          verification: {
            status: 'unverified',
          },
        },
      ],
      web3_wallets: [],
      external_accounts: [],
    } as unknown as UserJSON);

    expect(userWithUnverifiedPhone.hasVerifiedPhoneNumber).toEqual(false);

    const userWithVerifiedPhone = new User({
      email_addresses: [],
      phone_numbers: [
        {
          phoneNumber: '+306900000000',
          verification: {
            status: 'verified',
          },
        },
      ],
      web3_wallets: [],
      external_accounts: [],
    } as unknown as UserJSON);

    expect(userWithVerifiedPhone.hasVerifiedPhoneNumber).toEqual(true);
  });

  it('creates a new TOTP secret for the user', async () => {
    const totpJSON = {
      id: 'totp_foo',
      object: 'totp',
      secret: 'BANANA',
      uri: 'otpauth://totp/Cyberdyne%20%28development%29:vin@diesel.family?algorithm=SHA1&digits=6&issuer=Cyberdyne+%28development%29&period=30&secret=BANANA',
      verified: false,
      createdAt: 1659607590792,
      updatedAt: 1659607590792,
    };

    // @ts-ignore
    BaseResource._fetch = vi.fn().mockReturnValue(Promise.resolve({ response: totpJSON }));

    const user = new User({
      email_addresses: [],
      phone_numbers: [],
      web3_wallets: [],
      external_accounts: [],
    } as unknown as UserJSON);

    await user.createTOTP();

    // @ts-ignore
    expect(BaseResource._fetch).toHaveBeenCalledWith({
      method: 'POST',
      path: '/me/totp',
    });
  });

  it('verifies a TOTP secret', async () => {
    const totpJSON = {
      id: 'totp_foo',
      object: 'totp',
      verified: true,
      createdAt: 1659607590792,
      updatedAt: 1659607590792,
    };

    // @ts-ignore
    BaseResource._fetch = vi.fn().mockReturnValue(Promise.resolve({ response: totpJSON }));

    const user = new User({
      email_addresses: [],
      phone_numbers: [],
      web3_wallets: [],
      external_accounts: [],
    } as unknown as UserJSON);

    await user.verifyTOTP({ code: '123456' });

    // @ts-ignore
    expect(BaseResource._fetch).toHaveBeenCalledWith({
      method: 'POST',
      path: '/me/totp/attempt_verification',
      body: { code: '123456' },
    });
  });

  it('disables TOTP for a user', async () => {
    const deletedObjectJSON = {
      object: 'externaltotp_account',
      id: 'totp_foo',
      deleted: true,
    };

    // @ts-ignore
    BaseResource._fetch = vi.fn().mockReturnValue(Promise.resolve({ response: deletedObjectJSON }));

    const user = new User({
      email_addresses: [],
      phone_numbers: [],
      web3_wallets: [],
      external_accounts: [],
    } as unknown as UserJSON);

    await user.disableTOTP();

    // @ts-ignore
    expect(BaseResource._fetch).toHaveBeenCalledWith({
      method: 'DELETE',
      path: '/me/totp',
    });
  });

  it('creates backup codes', async () => {
    const backupCodeJSON = {
      object: 'backup_code',
      id: 'bcode_1234',
      codes: ['1234', '5678'],
    };

    // @ts-ignore
    BaseResource._fetch = vi.fn().mockReturnValue(Promise.resolve({ response: backupCodeJSON }));

    const user = new User({
      email_addresses: [],
      phone_numbers: [],
      web3_wallets: [],
      external_accounts: [],
    } as unknown as UserJSON);

    await user.createBackupCode();

    // @ts-ignore
    expect(BaseResource._fetch).toHaveBeenCalledWith({
      method: 'POST',
      path: '/me/backup_codes/',
    });
  });

  describe('Set the right fullName', () => {
    const cases: Array<[string | null | undefined, string | null | undefined, string | null | undefined]> = [
      // firstName, lastName, fullName
      ['A', 'B', 'A B'],
      ['', 'B', 'B'],
      ['A', '', 'A'],
      ['', '', null],
      [null, '', null],
      [null, null, null],
      [undefined, undefined, null],
    ];

    it.each(cases)("firstName: '%s', lastName: '%s'  =>  fullName: '%s'", (first_name, last_name, fullName) => {
      const user = new User({
        first_name,
        last_name,
      } as unknown as UserJSON);
      expect(user.fullName).toBe(fullName);
    });
  });

  it('.updatePassword triggers a request to change password', async () => {
    // @ts-ignore
    BaseResource._fetch = vi.fn().mockReturnValue(Promise.resolve({ response: {} }));

    const user = new User({} as unknown as UserJSON);
    const params = {
      currentPassword: 'current-password',
      newPassword: 'new-password',
    };
    await user.updatePassword(params);

    // @ts-ignore
    expect(BaseResource._fetch).toHaveBeenCalledWith({
      method: 'POST',
      path: '/me/change_password',
      body: params,
    });
  });

  it('.removePassword triggers a request to remove password', async () => {
    // @ts-ignore
    BaseResource._fetch = vi.fn().mockReturnValue(Promise.resolve({ response: {} }));

    const user = new User({} as unknown as UserJSON);
    const params = {
      currentPassword: 'current-password',
    };
    await user.removePassword(params);

    // @ts-ignore
    expect(BaseResource._fetch).toHaveBeenCalledWith({
      method: 'POST',
      path: '/me/remove_password',
      body: params,
    });
  });
});

describe('User Snapshots', () => {
  it('should match snapshot for user instance structure', () => {
    const userJSON: UserJSON = {
      object: 'user',
      id: 'user_123',
      first_name: 'John',
      last_name: 'Doe',
      username: 'johndoe',
      image_url: 'https://example.com/avatar.jpg',
      has_image: true,
      primary_email_address_id: 'email_123',
      primary_phone_number_id: null,
      primary_web3_wallet_id: null,
      email_addresses: [
        {
          object: 'email_address',
          id: 'email_123',
          email_address: 'john@example.com',
          verification: {
            object: 'verification',
            id: 'verification_123',
            status: 'verified',
            strategy: 'email_code',
            attempts: 1,
            expire_at: 1735689600000,
            verified_at_client: null,
            error: null,
          },
          linked_to: [],
          created_at: 1735689500000,
          updated_at: 1735689600000,
        },
      ],
      phone_numbers: [],
      web3_wallets: [],
      external_accounts: [],
      passkeys: [],
      saml_accounts: [],
      public_metadata: { role: 'user' },
      unsafe_metadata: { preferences: { theme: 'dark' } },
      created_at: 1735689500000,
      updated_at: 1735689600000,
      last_sign_in_at: 1735689550000,
      password_enabled: true,
      totp_enabled: false,
      backup_code_enabled: false,
      two_factor_enabled: false,
      banned: false,
      locked: false,
      lockout_expires_in_seconds: null,
      verification_attempts_remaining: 3,
      delete_self_enabled: true,
      create_organization_enabled: true,
      last_active_at: 1735689600000,
      profile_image_url: 'https://example.com/avatar.jpg',
      organization_memberships: [],
    };

    const user = new User(userJSON);

    const userSnapshot = {
      id: user.id,
      firstName: user.firstName,
      lastName: user.lastName,
      fullName: user.fullName,
      username: user.username,
      imageUrl: user.imageUrl,
      hasImage: user.hasImage,
      primaryEmailAddressId: user.primaryEmailAddressId,
      primaryPhoneNumberId: user.primaryPhoneNumberId,
      primaryWeb3WalletId: user.primaryWeb3WalletId,
      passwordEnabled: user.passwordEnabled,
      totpEnabled: user.totpEnabled,
      backupCodeEnabled: user.backupCodeEnabled,
      twoFactorEnabled: user.twoFactorEnabled,
      banned: user.banned,
      locked: user.locked,
      createdAt: user.createdAt?.getTime(),
      updatedAt: user.updatedAt?.getTime(),
      lastSignInAt: user.lastSignInAt?.getTime(),
      lastActiveAt: user.lastActiveAt?.getTime(),
      publicMetadata: user.publicMetadata,
      unsafeMetadata: user.unsafeMetadata,
      emailAddresses: user.emailAddresses.map(email => ({
        id: email.id,
        emailAddress: email.emailAddress,
        verification: email.verification
          ? {
              status: email.verification.status,
              strategy: email.verification.strategy,
            }
          : null,
      })),
    };

    expect(userSnapshot).toMatchSnapshot();
  });

  it('should match snapshot for minimal user', () => {
    const userJSON: UserJSON = {
      object: 'user',
      id: 'user_minimal',
      first_name: null,
      last_name: null,
      username: null,
      image_url: null,
      has_image: false,
      primary_email_address_id: null,
      primary_phone_number_id: null,
      primary_web3_wallet_id: null,
      email_addresses: [],
      phone_numbers: [],
      web3_wallets: [],
      external_accounts: [],
      passkeys: [],
      saml_accounts: [],
      public_metadata: {},
      unsafe_metadata: {},
      created_at: 1735689600000,
      updated_at: 1735689600000,
      last_sign_in_at: null,
      password_enabled: false,
      totp_enabled: false,
      backup_code_enabled: false,
      two_factor_enabled: false,
      banned: false,
      locked: false,
      lockout_expires_in_seconds: null,
      verification_attempts_remaining: 3,
      delete_self_enabled: true,
      create_organization_enabled: false,
      last_active_at: null,
      profile_image_url: null,
      organization_memberships: [],
    };

    const user = new User(userJSON);

    const userSnapshot = {
      id: user.id,
      firstName: user.firstName,
      lastName: user.lastName,
      fullName: user.fullName,
      username: user.username,
      imageUrl: user.imageUrl,
      hasImage: user.hasImage,
      passwordEnabled: user.passwordEnabled,
      totpEnabled: user.totpEnabled,
      backupCodeEnabled: user.backupCodeEnabled,
      twoFactorEnabled: user.twoFactorEnabled,
      banned: user.banned,
      locked: user.locked,
      createdAt: user.createdAt?.getTime(),
      updatedAt: user.updatedAt?.getTime(),
      lastSignInAt: user.lastSignInAt?.getTime(),
      lastActiveAt: user.lastActiveAt?.getTime(),
      publicMetadata: user.publicMetadata,
      unsafeMetadata: user.unsafeMetadata,
    };

    expect(userSnapshot).toMatchSnapshot();
  });
});
