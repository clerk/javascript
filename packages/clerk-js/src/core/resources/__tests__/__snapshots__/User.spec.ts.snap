// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`User Snapshots > should match snapshot for minimal user 1`] = `
{
  "backupCodeEnabled": false,
  "banned": undefined,
  "createdAt": 1735689600000,
  "firstName": null,
  "fullName": null,
  "hasImage": false,
  "id": "user_minimal",
  "imageUrl": "",
  "lastActiveAt": undefined,
  "lastName": null,
  "lastSignInAt": undefined,
  "locked": undefined,
  "passwordEnabled": false,
  "publicMetadata": {},
  "totpEnabled": false,
  "twoFactorEnabled": false,
  "unsafeMetadata": {},
  "updatedAt": 1735689600000,
  "username": null,
}
`;

exports[`User Snapshots > should match snapshot for user instance structure 1`] = `
{
  "backupCodeEnabled": false,
  "banned": undefined,
  "createdAt": 1735689500000,
  "emailAddresses": [
    {
      "emailAddress": "john@example.com",
      "id": "email_123",
      "verification": {
        "status": "verified",
        "strategy": "email_code",
      },
    },
  ],
  "firstName": "John",
  "fullName": "John Doe",
  "hasImage": true,
  "id": "user_123",
  "imageUrl": "https://example.com/avatar.jpg",
  "lastActiveAt": undefined,
  "lastName": "Doe",
  "lastSignInAt": 1735689550000,
  "locked": undefined,
  "passwordEnabled": true,
  "primaryEmailAddressId": "email_123",
  "primaryPhoneNumberId": null,
  "primaryWeb3WalletId": null,
  "publicMetadata": {
    "role": "user",
  },
  "totpEnabled": false,
  "twoFactorEnabled": false,
  "unsafeMetadata": {
    "preferences": {
      "theme": "dark",
    },
  },
  "updatedAt": 1735689600000,
  "username": "johndoe",
}
`;
